src/summarycode/file_cat.py

```diff
@@ -0,0 +1,1265 @@
+#
+# Copyright (c) nexB Inc. and others. All rights reserved.
+# ScanCode is a trademark of nexB Inc.
+# SPDX-License-Identifier: Apache-2.0
+# See http://www.apache.org/licenses/LICENSE-2.0 for the license text.
+# See https://github.com/nexB/scancode-toolkit for support or download.
+# See https://aboutcode.org for more information about nexB OSS projects.
+#
+
+from commoncode.datautils import String
+from plugincode.post_scan import PostScanPlugin
+from plugincode.post_scan import post_scan_impl
+from commoncode.cliutils import PluggableCommandLineOption
+from commoncode.cliutils import POST_SCAN_GROUP
+
+"""
+Categorize files.
+"""
+
+# Tracing flag
+TRACE = False
+
+
+def logger_debug(*args):
+    pass
+
+
+if TRACE:
+    import logging
+    import click
+
+    class ClickHandler(logging.Handler):
+        _use_stderr = True
+
+        def emit(self, record):
+            try:
+                msg = self.format(record)
+                click.echo(msg, err=self._use_stderr)
+            except Exception:
+                self.handleError(record)
+
+    logger = logging.getLogger(__name__)
+    logger.handlers = [ClickHandler()]
+    logger.propagate = False
+    logger.setLevel(logging.DEBUG)
+
+    def logger_debug(*args):
+        return logger.debug(" ".join(isinstance(a, str) and a or repr(a) for a in args))
+
+
+@post_scan_impl
+class FileCategorizer(PostScanPlugin):
+    """
+    Categorize a file.
+    """
+
+    resource_attributes = dict(
+        [
+            ("analysis_priority", String(help="Analysis priority.")),
+            ("file_category", String(help="File category.")),
+            ("file_subcategory", String(help="File subcategory.")),
+        ]
+    )
+
+    sort_order = 50
+
+    options = [
+        PluggableCommandLineOption(
+            ("--file-cat",),
+            is_flag=True,
+            default=False,
+            help="Categorize files.",
+            help_group=POST_SCAN_GROUP,
+            sort_order=50,
+        )
+    ]
+
+    def is_enabled(self, file_cat, **kwargs):
+        return file_cat
+
+    def process_codebase(self, codebase, file_cat, **kwargs):
+        if not file_cat:
+            return
+
+        for resource in codebase.walk(topdown=True):
+            category = categorize_resource(resource)
+            if not category:
+                continue
+            resource.analysis_priority = category.analysis_priority
+            resource.file_category = category.file_category
+            resource.file_subcategory = category.file_subcategory
+            resource.save(codebase)
+
+
+class Categorizer:
```

Comment:
Add some docstring to document what is this class used for and how to extend it
```suggestion
class Categorizer:
    """
    A categorizer can categorize a file Resource based on its path, name,
    extension and other attributes such as file or mime type.
    Subclasses must override the ``categorize`` class method.
    """
```